Enable EKS upgrade test

pydctw opened the pull request:
**What type of PR is this?**


**What this PR does / why we need it**:
Enable EKS upgrade test, updated k8s version to EKS supported versions and setup AWSClusterStaticIdentity in a central location for use by multiple tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3228 

**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [x] adds or updates e2e tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. 
2. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE"....however we encourage contributors to never use this as release notes are incredible useful.
-->
```release-note
NONE
```

pydctw commented:
/test

pydctw commented:
/test pull-cluster-api-provider-aws-e2e-eks

pydctw commented:
/hold

pydctw commented:
The upgrade test actually passed.
```
• [SLOW TEST:2638.881 seconds]
EKS Cluster upgrade test
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/test/e2e/suites/managed/upgrade_test.go:37
  [managed] [upgrade] should create a cluster and upgrade the kubernetes version
  /home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/test/e2e/shared/common.go:179
```

pydctw commented:
The failure is in the eks cluster test - it failed while trying to create a secret for AWSClusterStaticIdentity.
```
secrets \"e2e-account-creds\" already exists"
```

Are we supposed to set up AWSClusterStaticIdentity once and use for both upgrade and regular eks tests? I thought they should be set up per cluster but will do more investigation.

richardcase commented:
> Are we supposed to set up AWSClusterStaticIdentity once and use for both upgrade and regular eks tests? I thought they should be set up per cluster but will do more investigation.

They should be done once in the bootstrap cluster. So we probably need to move:

```go
ginkgo.By("setting up AWS static credentials")
shared.SetupStaticCredentials(ctx, namespace, e2eCtx)
```

To the shared test setup. So somewhere in [Node1BeforeSuite](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/shared/suite.go#L59) when the bootstrap cluster has been created.

We will also need to move the credential cleanup:

```go
ginkgo.By("Deleting AWS static credentials")
shared.CleanupStaticCredentials(ctx, namespace, e2eCtx)
```

to the shared test teardown. So somewhere in [Node1AfterSuite](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/shared/suite.go#L256).

pydctw commented:
/test pull-cluster-api-provider-aws-e2e-eks

pydctw commented:
@richardcase, thanks for sharing the idea. I have moved SetupStaticCredentials and CleanupStaticCredentials functions to suite.go so that it can be set up once and used for both tests.

Note that I refactored AWSClusterStaticIdentity functions as part of this.
* Not to specify namespace for the object as it is a cluster-scoped object. 
* Allow all namespaces (AllowedNamespaces) so that it can be used for each test run with a newly created namespace.

pydctw commented:
/hold cancel

sedefsavas commented on test/e2e/shared/suite.go:
```diff
@@ -169,6 +169,9 @@ func Node1BeforeSuite(e2eCtx *E2EContext) []byte {
 	CreateAWSClusterControllerIdentity(e2eCtx.Environment.BootstrapClusterProxy.GetClient())
 
+	By("Setting up AWS static credentials")
+	SetupStaticCredentials(e2eCtx)
```
Do we need to create static credentials for unmanaged tests?

richardcase replied:
This should do it for both the managed and unmanaged

sedefsavas replied:
We are using AWSClusterControllerIdentity in unmanaged clusters, should not need this.

richardcase replied:
Sorry @sedefsavas - i misread that

pydctw replied:
This doesn't affect any test unless the test uses AWSClusterStaticIdentity. We can set it up for only managed test suite but someone may want to use AWSClusterStaticIdentity for unmanaged suite in the future.

pydctw replied:
Just saw new comments :) Either way is fine. Let me know what you think.

sedefsavas replied:
We can add appropriate identity creation to specific tests like we do in nested multi-tenancy tests.
I'd opt for not creating unused objects that may or may not be used in the future.

richardcase replied:
Ok, so moving this back into the managed suite then, where it can be called from both tests

pydctw replied:
Not moving back to the managed suite but added conditions for the setup and deletion
```
	if e2eCtx.IsManaged {
		By("Setting up AWS static credentials")
		SetupStaticCredentials(e2eCtx)
	}
```

pydctw replied:
I think we may want to split suite.go for managed and unmanaged in the long term as the needs are different.

richardcase commented on test/e2e/shared/identity.go:
```diff
@@ -60,38 +59,33 @@ func SetupStaticCredentials(ctx context.Context, namespace *corev1.Namespace, e2
-				AllowedNamespaces: &infrav1.AllowedNamespaces{
-					NamespaceList: []string{namespace.Name},
-				},
+				AllowedNamespaces: &infrav1.AllowedNamespaces{},
```
I think it would be better to keep it limited to the namespace of the test

pydctw replied:
I debated about it. If we want to use the namespace of the test which is created at the start of the each test run, it will be better to move back SetupStaticCredentials to each test as we don't know the test namespace in advance. 
Any thoughts?

richardcase replied:
Ah yes, good point @pydctw . Ignore me :smile:

pydctw commented:
```
I0301 16:22:49.605924   22251 logr.go:252]  "msg"="Wait completed, proceeding to shutdown the manager"  
panic: Failed to stop envtest: timeout waiting for process kube-apiserver to stop
```
Looks likes a flake with envtest. Rerunning the test

/retest

sedefsavas commented:
/hold

until above comment is addressed.

pydctw commented:
Looks like the CI env is slow or sth today. Failed at verify, which has never failed for the PR before.
```
Init container clonerefs not ready: (state: waiting, reason: "PodInitializing", message: "") Init container initupload not ready: (state: waiting, reason: "PodInitializing", message: "") Init container place-entrypoint not ready: (state: waiting, reason: "PodInitializing", message: "")
```

/retest

pydctw commented:
/test pull-cluster-api-provider-aws-e2e-eks

pydctw commented:
/pull-cluster-api-provider-aws-verify

pydctw commented:
/test pull-cluster-api-provider-aws-verify

pydctw commented:
Having an issue with the env today. pull-cluster-api-provider-aws-verify continues to fail. Will try again later.
```
Init container clonerefs not ready: (state: waiting, reason: "PodInitializing", message: "") Init container initupload not ready: (state: waiting, reason: "PodInitializing", message: "") Init container place-entrypoint not ready: (state: waiting, reason: "PodInitializing", message: "")
```

sedefsavas commented:
/lgtm

pydctw commented:
All tests passed.

/hold cancel
